fdbserver/GrvProxyServer.actor.cpp

```diff
@@ -302,6 +304,133 @@ ACTOR Future<Void> healthMetricsRequestServer(GrvProxyInterface grvProxy,
+				state Optional<Value> sizeLimit =
+				    wait(tr->get(Key("\xff\x02/fdbClientInfo/client_txn_size_limit/"_sr)));
+
+				tr->setOption(FDBTransactionOptions::SPECIAL_KEY_SPACE_ENABLE_WRITES);
+				// The value doesn't matter too much, as long as the key is set.
+				tr->set(migratedKey.contents(), "1"_sr);
+				if (sampleRate.present()) {
+					const double sampleRateDbl =
+					    BinaryReader::fromStringRef<double>(sampleRate.get().contents(), Unversioned());
+					Tuple rate = Tuple().appendDouble(sampleRateDbl);
+					tr->set(GlobalConfig::prefixedKey(fdbClientInfoTxnSampleRate), rate.pack());
+				}
+				if (sizeLimit.present()) {
+					const int64_t sizeLimitInt =
+					    BinaryReader::fromStringRef<int64_t>(sizeLimit.get().contents(), Unversioned());
+					Tuple size = Tuple().append(sizeLimitInt);
+					tr->set(GlobalConfig::prefixedKey(fdbClientInfoTxnSizeLimit), size.pack());
+				}
+
+				wait(tr->commit());
+				break;
+			} catch (Error& e) {
+				// If multiple fdbserver processes are started at once, they will all
+				// attempt this migration at the same time, sometimes resulting in
+				// aborts due to conflicts. Purposefully avoid retrying, making this
+				// migration best-effort.
+				TraceEvent(SevInfo, "GlobalConfig_RetryableMigrationError").errorUnsuppressed(e).suppressFor(1.0);
+				wait(tr->onError(e));
+				tr.clear();
+				// tr is cleared, so it won't backoff properly. Use custom backoff logic here.
+				wait(backoff.onError());
+			}
+		}
+	} catch (Error& e) {
+		// Catch non-retryable errors (and do nothing).
+		TraceEvent(SevWarnAlways, "GlobalConfig_MigrationError").error(e);
+	}
+	req.reply.send(Void());
+	return Void();
+}
+
+ACTOR Future<Void> globalConfigHandleRefreshBatch(std::vector<GlobalConfigRefreshRequest> requests,
+                                                  GrvProxyData* grvProxyData) {
+	if (requests.size() == 0) {
+		return Void();
+	}
+
+	state Backoff backoff;
+	state Reference<ReadYourWritesTransaction> tr;
+	loop {
+		try {
+			tr = makeReference<ReadYourWritesTransaction>(grvProxyData->cx);
+			tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+			RangeResult result = wait(tr->getRange(globalConfigDataKeys, CLIENT_KNOBS->TOO_MANY));
+			for (const auto& req : requests) {
+				req.reply.send(GlobalConfigRefreshReply{ result });
+			}
+			break;
+		} catch (Error& e) {
+			wait(tr->onError(e));
+			tr.clear();
+			// tr is cleared, so it won't backoff properly. Use custom backoff logic here.
+			wait(backoff.onError());
+		}
+	}
+	return Void();
+}
+
+ACTOR Future<Void> globalConfigRefreshBatcher(GrvProxyData* grvProxyData,
+                                              GrvProxyInterface grvProxy,
+                                              PromiseStream<std::vector<GlobalConfigRefreshRequest>> out) {
+	loop {
+		state Future<Void> timeout = delay(SERVER_KNOBS->GLOBAL_CONFIG_MAX_REFRESH_BATCH_INTERVAL);
```

Comment:
I think this batching approach works fine. As another option, we could also flip it a bit and do the DB read periodically, cache the result, and reply to each request immediately with the latest cached data. That approach may support longer intervals between reads while still keeping response times low to requests, but I don't actually have a strong preference.

Reply:
I guess one interesting thing that could happen in this case is that if the database is down or struggling, you will accumulate batches as more and more requests come in. If the number of clients is constant, then this will eventually stop once every client has a request outstanding. If we enter some scenario where clients are starting and terminating repeatedly (perhaps because of the down cluster), then I suppose you could accumulate an unbounded number of requests.

Each of the batches will run its own transaction in a loop simultaneously, and if the problem is that the cluster is struggling rather than completely down, this could potentially lead to extra traffic that makes things worse. Or maybe if the cluster is down, you get a sudden spike of heavy reads when the cluster does become available again.

In the other model, there is only one actor that is reading the database in a loop, so if it gets stuck or retries a bunch, the traffic there is still small. You also can still reply to clients even without being able to read, so the requests shouldn't accumulate.

Reply:
Since `GlobalConfig` is tied to a database now, you could even run into this problem with a fixed number of client in the case where a single client is repeatedly opening a database.

Reply:
One positive current aspect is the second delay between creating batches. We could also add something like a delay after processing a batch (of a tenth of a second or something). This would prevent spamming the database with a bunch of calls if the request queue is large, but would still eventually clear the queue.

Reply:
I think there are ways we could manage the batches if need be, but it might still be simpler to adopt the model where we have only one actor reading and each request immediately returns the latest cached result. In that model we only have one reader and no queued requests.

Reply:
I just implemented this but I think there is a problem. The client realizes it needs to refresh its state and calls into the GRV Proxy. But if the proxy has the old result cached and returns that, the client may think there weren't any changes and will return as such. Then future reads the client does don't cause another refresh, meaning the client is permanently out of date. The issue is that calls to refresh need to return a consistent value at the time of the request, so returning potentially stale data from a cache won't work here I don't think.

Reply:
One alternative implementation to creating batches every second is to have a single batch that all incoming requests get added to. Then, when the database is able to run the read and gets a result, it can reply to everyone in the batch, then clear the batch and allow everything to restart. This will maintain consistency but still result in only one database read no matter how slow it ends up being.

Reply:
I see, I was thinking that the refresh happened by polling, but it happens upon receiving a signal that something has changed from the coordinator. I think this could still work, but you would need to communicate some details about the version in the exchange with the proxy and do some polling.

A third option here is to have the proxy perform reads only in response to the same signal in `ClientDBInfo` or perhaps if it gets a request for a newer version from a client. Then you could accumulate requests based on the version and fulfill them all when you have it. This would require some mechanism to limit the total requests in queue if things get stuck.

Reply:
> you would need to communicate some details about the version in the exchange with the proxy and then repeatedly poll the proxy with a backoff

The question is if it is better to have more messages between clients and the proxy vs having the proxy do more reads of the database. I think if we try to reduce the number of times it reads from the system key range, limiting it to once a second or so instead of once per client, we could get a similar effect.

It may also be possible to avoid polling by including the last update version in the RefreshRequest and only sending a reply from the proxy when the read version is larger. But then you may still run into consistency issues. I'm not sure if this situation is possible, but I can imagine something like:
* client is on version 6
* different client does a write at version 7
* GRV proxy updates its cached state
* different client does a write at version 8
* original client receives updated ClientDBInfo, decides to refresh its state, reads newer but not fully up to date state at version 7.
Now the client has updated state but not fully up to date. I think we always want to kick off a read at a time later than when the proxy receives the refresh request in order to guarantee we are returning the latest state.

>A third option here is to have the proxy perform reads only in response to the same signal in ClientDBInfo or perhaps if it gets a request for a newer version from a client

Another worry here is if the `ClientDBInfo` syncs faster to the client than to the proxy, the proxy may not have received the signal that it needs to update before responding to the client.

Reply:
This wouldn't use more messages, but it would include some version details in the existing ones. For example, the proxy could communicate back what version it had read, and if the client was expecting something newer it could try again (this would be the polling with backoff approach). Or in the third option I suggested, the client might send what version it is looking for and the proxy would respond once it had that version available.

Reply:
> I think we always want to kick off a read at a time later than when the proxy receives the refresh request in order to guarantee we are returning the latest state.

This is a good way to accomplish something similar to the version-based response on the proxy. Keep in mind, though, that if you do either of those things we should probably still track how many outstanding requests there are and start rejecting them if there are too many. Then the clients will probably have to start polling.

Reply:
I implemented a version where the client asks for the latest version included in the `ClientDBInfo` update. Can you look at the backoff method I used -- I wasn't quite sure what to use since the client doesn't explicitly use a transaction object.